pkg/codegen/internal/tstypes/tstypes.go

```diff
@@ -0,0 +1,222 @@
+// inserted parentheses.
+func TypeLiteral(ast TypeAst) string {
+	tokens := (&typeScriptTypeUnparser{}).unparse(ast)
+	return toLiteral(tokens)
+}
+
+// Builds a type identifier (possibly qualified such as
+// "my.module.MyType") or a primitive such as "boolean".
+func Identifier(id string) TypeAst {
+	return &idType{id}
+}
+
+// Builds a `T[]` type from a `T` type.
+func Array(t TypeAst) TypeAst {
+	return &arrayType{t}
+}
+
+// Builds a `{[key: string]: T}` type from a `T` type.
+func StringMap(t TypeAst) TypeAst {
+	return &mapType{t}
+}
+
+// Builds a union `A | B | C` type.
+func Union(t ...TypeAst) TypeAst {
+	if len(t) == 0 {
+		panic("At least one type is needed to form a Union, none are given")
+	}
+	if len(t) == 1 {
+		return t[0]
+	}
+	return &unionType{t[0], t[1], t[2:]}
+}
+
+// Normalizes by unnesting unions `A | (B | C) => A | B | C`.
+func Normalize(ast TypeAst) TypeAst {
+	return transform(ast, func(t TypeAst) TypeAst {
+		switch v := t.(type) {
+		case *unionType:
+			var all []TypeAst
+			for _, e := range v.all() {
+				switch ev := e.(type) {
+				case *unionType:
+					all = append(all, ev.all()...)
+				default:
+					all = append(all, ev)
+				}
+			}
+			return Union(all...)
+		default:
+			return t
+		}
+	})
+}
+
+func transform(t TypeAst, f func(x TypeAst) TypeAst) TypeAst {
+	switch v := t.(type) {
+	case *unionType:
+		var ts []TypeAst
+		for _, x := range v.all() {
+			ts = append(ts, transform(x, f))
+		}
+		return f(Union(ts...))
+	case *arrayType:
+		return f(&arrayType{transform(v.arrayElement, f)})
+	case *mapType:
+		return f(&mapType{transform(v.mapElement, f)})
+	default:
+		return f(t)
+	}
+}
+
+type idType struct {
+	id string
+}
+
+func (*idType) depth() int {
+	return 1
+}
+
+var _ TypeAst = &idType{}
+
+type mapType struct {
+	mapElement TypeAst
+}
+
+func (t *mapType) depth() int {
+	return t.mapElement.depth() + 1
+}
+
+var _ TypeAst = &mapType{}
+
+type arrayType struct {
+	arrayElement TypeAst
+}
+
+func (t *arrayType) depth() int {
+	return t.arrayElement.depth() + 1
+}
+
+var _ TypeAst = &arrayType{}
+
+type unionType struct {
+	t1    TypeAst
+	t2    TypeAst
+	tRest []TypeAst
+}
+
+func (t *unionType) all() []TypeAst {
+	return append([]TypeAst{t.t1, t.t2}, t.tRest...)
+}
+
+func (t *unionType) depth() int {
+	var maxDepth = 0
+	for _, t := range t.all() {
+		d := t.depth()
+		if d > maxDepth {
+			maxDepth = d
+		}
+	}
+	return maxDepth
+}
+
+var _ TypeAst = &unionType{}
+
+type typeTokenKind string
+
+const (
+	openParen  typeTokenKind = "("
+	closeParen               = ")"
+	openMap                  = "{[key: string]: "
+	closeMap                 = "}"
+	identifier               = "x"
+	array                    = "[]"
+	union                    = " | "
+)
+
+type typeToken struct {
+	kind  typeTokenKind
+	value string
+}
+
+type typeScriptTypeUnparser struct{}
+
+func (u *typeScriptTypeUnparser) unparse(ast TypeAst) []typeToken {
```

Comment:
are these methods just for scoping? if so, these can take a value receiver:

```suggestion
func (u typeScriptTypeUnparser) unparse(ast TypeAst) []typeToken {
```